### ISSUE TITLE
SD Card Corruption using extern "C"

### ISSUE THREAD

**Issue by wildnut** [espressif/arduino-esp32#591]
### Hardware:
Board:							AnalogLamb ESP32
Core Installation/update date:			Latest
IDE name:							Eclipse

### Description:
Hi,

First of all, thanks for the great library!  :)  It certainly makes many tasks much easier.  

So here's the problem I'm having and perhaps I am doing something wrong.  In Platform IO, the SD card library works great.  I can read/write with no problems using the great steps provided in the example and SD image.   No problems with PlatformIO using the arduino framework even after rebooting, taking the card out, etc.   All is well!

However, I have started recently using Eclipse and the "make" command.  I copied the exact same code that's in the SD example, changed void setup() to extern "C" void app_main() and then I upload the new code (I don't change the setting in make menuconfig to use Arduino, I leave it alone) using the make command from the command line.  Everything uploads fine and the program runs.  Here's a portion:   

Writing file: /hello.txt
File written
Appending to file: /hello.txt
Message appended
Reading file: /hello.txt
Read from file: Hello World!
Deleting file: /foo.txt
File deleted
Renaming file /hello.txt to /foo.txt
File renamed
Reading file: /foo.txt
Read from file: Hello World!
0 bytes read for 0 ms

It stops with the last line above.  

The "foo.txt" file was a carry-over from the PlatformIO test which gets deleted per the example sketch.   So, I then reboot the ESP32 and the card will no longer mount.  I then try the card in Windows and it can no longer be read nor can it be formatted.  I've tried using some of them in Android and a camera and the cards appear to be no longer usable.   This is my 4th SD card that this has happened to . . . I've used 3 micro SD cards and one large one from different manufacturers and they all seem to get corrupted to the point that they can't be used at all.

Am I doing something wrong with my code?  As mentioned, I am only adding the extern "C" section and then re-uploading.   Has anyone else tried this process as outline above . . . using extern "C" rather than the Arduino menuconfig changes?

Thanks for your help!  Once again, great library . . . but I've hit a bump in the road.  :)

**Comment by wildnut** [comment]
An update to say that I also added "#include "Arduino.h"" at the very top which is the only addition besides the extern "C" line added.   As mentioned, PlatformIO works great - this corruption issue is only happening when trying to use the SD library as a component of ESP-IDF as outlined here:

https://github.com/espressif/arduino-esp32/blob/master/docs/esp-idf_component.md

Thanks!

**Comment by wildnut** [comment]
Well, at least so far, the issue has been fixed.  I first tried changing the make menuconfig to use Arduino and then using setup() and loop() - which worked fine without SD card corruption.  I then removed those settings and changed the menuconfig to use app_main instead of Arduino.   I then re-looked over my code and realized that I had left out "initArduino();" in the app_main which, at least for the time being, seems to be the resolution unless it happens again.  :)    I also forgot to mention in the previous message that I did include "#include "Arduino.h"" at the top.  So, from what I can tell thus far with a very limited amount of testing . . . "initArduino()" is the solution to preventing card corruption.  Time will tell though, but hopefully it's that simple.   So, it was a mistake on my part.  Surprise!  Surprise!!  

Thanks again for all your hard work in creating such an amazing library.
